internal/vm/vsock.go

```diff
@@ -0,0 +1,138 @@
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package vm
+
+import (
+	"context"
+	"io"
+	"net"
+	"time"
+
+	"github.com/mdlayher/vsock"
+	"github.com/sirupsen/logrus"
+)
+
+// VSockDial attempts to connect to a vsock listener at the provided cid and port with a hardcoded number
+// of retries.
+func VSockDial(reqCtx context.Context, logger *logrus.Entry, contextID, port uint32) (net.Conn, error) {
+	// VM should start within 200ms, vsock dial will make retries at 100ms, 200ms, 400ms, 800ms, 1.6s, 3.2s, 6.4s
+	const (
+		retryCount      = 7
+		initialDelay    = 100 * time.Millisecond
+		delayMultiplier = 2
+	)
+
+	var lastErr error
+	var currentDelay = initialDelay
+
+	for i := 1; i <= retryCount; i++ {
+		select {
+		case <-reqCtx.Done():
+			return nil, reqCtx.Err()
+		default:
+			conn, err := vsock.Dial(contextID, port)
+			if err == nil {
+				logger.WithField("connection", conn).Debug("Dial succeeded")
+				return conn, nil
+			}
+
+			logger.WithError(err).Warnf("vsock dial failed (attempt %d of %d), will retry in %s", i, retryCount, currentDelay)
+			time.Sleep(currentDelay)
+
+			lastErr = err
+			currentDelay *= delayMultiplier
+		}
+	}
+
+	logger.WithError(lastErr).WithFields(logrus.Fields{"context_id": contextID, "port": port}).Error("vsock dial failed")
+	return nil, lastErr
+}
+
+// VSockDialConnector provides an IOConnector interface to the VSockDial function.
+func VSockDialConnector(contextID, port uint32) IOConnector {
+	return func(procCtx context.Context, logger *logrus.Entry) (<-chan io.ReadWriteCloser, <-chan error) {
+		returnCh := make(chan io.ReadWriteCloser)
+		errCh := make(chan error)
+		go func() {
+			defer close(returnCh)
+			defer close(errCh)
+
+			conn, err := VSockDial(procCtx, logger, contextID, port)
+			if err != nil {
+				errCh <- err
+			} else {
+				returnCh <- conn
+			}
+		}()
+
+		return returnCh, errCh
+	}
+}
+
+// VSockAcceptConnector provides an IOConnector that establishes the connection by listening on the provided
+// vsock port and accepting the first connection that comes in.
+func VSockAcceptConnector(port uint32) IOConnector {
+	return func(procCtx context.Context, logger *logrus.Entry) (<-chan io.ReadWriteCloser, <-chan error) {
+		returnCh := make(chan io.ReadWriteCloser)
+		errCh := make(chan error)
+
+		go func() {
+			defer close(returnCh)
+			defer close(errCh)
+
+			listener, err := vsock.Listen(port)
+			if err != nil {
+				errCh <- err
+				return
+			}
+
+			defer listener.Close()
+
+			for range time.NewTicker(10 * time.Millisecond).C {
+				select {
+				case <-procCtx.Done():
+					errCh <- procCtx.Err()
+					return
+				default:
+					// accept is non-blocking so try to accept until we get a connection
+					conn, err := listener.Accept()
+					if err == nil {
+						returnCh <- conn
+						return
+					}
+
+					if isTemporaryNetErr(err) {
```

Comment:
nit: I think would much prefer if we had the temporary logic stuff outside of the if statement or vice versa

```go
if isTemporaryNetErr(err) {
    logger.WithError(err).Debug("temporary stdio vsock accept failure")
    continue
}
   
logger.WithError(err).Error("non-temporary stdio vsock accept failure")
errCh <- err
return
```